pkg/kubelet/rocket/rocket.go

```diff
@@ -0,0 +1,453 @@
+	if c.Dir != "" {
+		result = append(result, fmt.Sprintf("--dir=%s", c.Dir))
+	}
+	return result
+}
+
+// New creats the rocket container runtime which implements the container runtime interface.
+// It will test if the 'rkt' binary is in the $PATH, and whether we can get the
+// version of it. If so, creates the rocket container runtime, other returns an error.
+func New(config *Config) (*Runtime, error) {
+	systemd, err := dbus.New()
+	if err != nil {
+		return nil, err
+	}
+	// Test if 'rkt' is in $PATH.
+	absPath, err := exec.LookPath(rocketBinName)
+	if err != nil {
+		return nil, err
+	}
+	rkt := &Runtime{
+		systemd: systemd,
+		absPath: absPath,
+		config:  config,
+	}
+
+	// Simply verify the binary by trying 'rkt version'.
+	result, err := rkt.Version()
+	if err != nil {
+		return nil, err
+	}
+	if _, found := result[rocketBinName]; !found {
+		return nil, fmt.Errorf("rocket: cannot get the version of rocket")
+	}
+	glog.V(4).Infof("Rocket version: %v.", result)
+	return rkt, nil
+}
+
+// Version invokes 'rkt version' to get the version information of the rocket
+// runtime on the machine.
+// The return values are a map of component:version.
+//
+// Example:
+// rkt:0.3.2+git
+// appc:0.3.0+git
+//
+func (r *Runtime) Version() (map[string]string, error) {
+	output, err := r.RunCommand("version")
+	if err != nil {
+		return nil, err
+	}
+
+	// Example output for 'rkt version':
+	// rkt version 0.3.2+git
+	// appc version 0.3.0+git
+	result := make(map[string]string)
+	for _, line := range output {
+		tuples := strings.Split(strings.TrimSpace(line), " ")
+		if len(tuples) != 3 {
+			glog.Warningf("Cannot parse the output: %q.", line)
+			continue
+		}
+		result[tuples[0]] = tuples[2]
+	}
+	return result, nil
+}
+
+// ListPods runs 'systemctl list-unit' and 'rkt list' to get the list of all the appcs.
+// Then it will use the result to contruct list of pods.
+func (r *Runtime) ListPods() ([]*api.Pod, error) {
+	glog.V(4).Infof("Rocket is listing pods.")
+
+	units, err := r.systemd.ListUnits()
+	if err != nil {
+		return nil, err
+	}
+
+	var pods []*api.Pod
+	for _, u := range units {
+		if strings.HasPrefix(u.Name, kubernetesUnitPrefix) {
+			pod, err := r.makePod(u.Name)
+			if err != nil {
+				glog.Warningf("Cannot construct pod from unit file: %v.", err)
+				continue
+			}
+			pods = append(pods, pod)
+		}
+	}
+	return pods, nil
+}
+
+// RunPod first creates the unit file for a pod, and then invokes
+// 'systemctl start' to run that unit file.
+func (r *Runtime) RunPod(pod *api.BoundPod, volumeMap map[string]volume.Interface) error {
+	glog.V(4).Infof("Rocket starts to run pod: name %q.", pod.Name)
+
+	name, needReload, err := r.preparePod(pod, volumeMap)
+	if err != nil {
+		return err
+	}
+	if needReload {
+		if err := r.systemd.Reload(); err != nil {
+			return err
+		}
+	}
+
+	ch := make(chan string)
+	_, err = r.systemd.StartUnit(name, "replace", ch)
+	if err != nil {
+		return err
+	}
+	if status := <-ch; status != "done" {
```

Comment:
@yifan-gu Well, `StartUnit` calls must reference a unit that already exists/is known by systemd (i.e. one you've written to disk), whereas with `StartTransientUnit` you can just pass in an entire unit and it's only ever hosted in memory.

Since we don't have any persistence needs (i.e. we're writing the unit files in /run anyway) we *might* be able to get away with using TransientUnits for everything but there's probably some subtle reason why not that I'm missing right now, so I think just stick with this approach.